File: backend-nestjs/src/controllers/public-sitemap.controller.ts
import { Controller, Get, Header, Param, Res } from '@nestjs/common';
import type { Response } from 'express';
import { SitemapService } from '../services/sitemap.service';

@Controller('/public/:tenant')
export class PublicSitemapController {
  constructor(private readonly sitemapService: SitemapService) {}

  @Get('sitemap.xml')
  @Header('Content-Type', 'application/xml; charset=utf-8')
  async sitemapIndex(@Param('tenant') tenant: string, @Res() response: Response): Promise<void> {
    const resolved = await this.sitemapService.getPublicSitemapXml(tenant);
    response.send(resolved.xml);
  }

  @Get('sitemap-:type-:index.xml')
  @Header('Content-Type', 'application/xml; charset=utf-8')
  async sitemapChunk(
    @Param('tenant') tenant: string,
    @Param('type') type: string,
    @Param('index') index: string,
    @Res() response: Response,
  ): Promise<void> {
    const resolved = await this.sitemapService.getPublicSitemapXml(tenant, type, Number(index));
    response.send(resolved.xml);
  }
}

File: backend/src/main/java/com/contentplatform/backend/interfaces/web/controller/PublicSitemapController.java
package com.contentplatform.backend.interfaces.web.controller;

import com.contentplatform.backend.application.service.SitemapService;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/public/{tenant}")
public class PublicSitemapController {
    private final SitemapService sitemapService;

    public PublicSitemapController(SitemapService sitemapService) {
        this.sitemapService = sitemapService;
    }

    @GetMapping(value = "/sitemap.xml", produces = MediaType.APPLICATION_XML_VALUE)
    public ResponseEntity<String> sitemap(@PathVariable String tenant) {
        return ResponseEntity.ok(sitemapService.getPublicSitemapXml(tenant, null, null));
    }

    @GetMapping(value = "/sitemap-{type}-{index}.xml", produces = MediaType.APPLICATION_XML_VALUE)
    public ResponseEntity<String> sitemapChunk(@PathVariable String tenant,
                                               @PathVariable String type,
                                               @PathVariable Integer index) {
        return ResponseEntity.ok(sitemapService.getPublicSitemapXml(tenant, type, index));
    }
}

